Port type system tests from CoreRT repo

nattress opened the pull request:
Port the managed type system unit tests and wire them up to `build.cmd -test` so they run if the `clr.tools` subset is included in compilation. We're missing USG support so those tests are disabled. Similarly, the tests for interfaces on arrays don't work in this repo since we never ported the `Array<T>` weirdo functionality.

The first commit is a simple xcopy, and subsequent commits make the tests work in this repo. Currently these are not enabled in CI and I could use some advice there since as I understand the plan is to add Pri 0 tests to `build -test` for the coreclr partition in the future. If we just have fast running unit tests then I'd be okay with these running as part of the product build when the coreclr partition has changes. Or we could build / run them as part of the test build.

Implements issue https://github.com/dotnet/runtime/issues/200

cc @dotnet/crossgen-contrib

nattress committed:
xcopy type system tests

nattress committed:
Tests build and pass

`build.cmd -test -s clr.tools -rc checked -a x64` builds the product and unit tests, and then runs the unit tests.

jkoritzinsky commented on eng/Subsets.props:
```diff
@@ -150,6 +150,7 @@
                              $(CoreClrProjectRoot)src\tools\dotnet-pgo\dotnet-pgo.csproj;
                              $(CoreClrProjectRoot)src\tools\r2rtest\R2RTest.csproj" Category="clr" BuildInParallel="true" />
     <ProjectToBuild Include="$(CoreClrProjectRoot)src\tools\aot\crossgen2\crossgen2.csproj" Category="clr" />
+    <ProjectToBuild Include="$(CoreClrProjectRoot)src\tools\aot\ILCompiler.TypeSystem.ReadyToRun.Tests\ILCompiler.TypeSystem.ReadyToRun.Tests.csproj" Test="true" Category="clr" />
```
I don't believe we need the `Test="true"` attribute here unless the intent is to only build this project when we're running the tests.

nattress replied:
It looks required; I tried removing it and the tests stopped running with `build.cmd -test`. With `Test="true"`, I still see the tests getting built without `-test` passed to build.

jkoritzinsky reviewed (commented):
Just one question from me. Other than that, if this works to run the tests locally then it looks good to me.

jkotas commented on src/coreclr/src/tools/aot/ILCompiler.TypeSystem.ReadyToRun.Tests/TestTypeSystemContext.cs:
```diff
@@ -57,21 +59,9 @@ public override ModuleDesc ResolveAssembly(System.Reflection.AssemblyName name,
             return _metadataFieldLayout;
         }
 
-        protected override RuntimeInterfacesAlgorithm GetRuntimeInterfacesAlgorithmForNonPointerArrayType(ArrayType type)
```
Put these under `#if !READYTORUN` instead of deleting them?

MichalStrehovsky replied:
We could also just pull in the `ArrayOfTRuntimeInterfacesAlgorithm.cs` file from the CoreRT repo, include it in the test project (no need to include from the shipping binary, I think), and it would let us run a couple more tests around array casting.

The algorithm and `Array<T>` is really just a way for the type system to learn what interfaces array has - it's not that important. What matters is that things like array casting work and they build on top of that.

MichalStrehovsky replied:
(And maybe we could do the same for the `UniversalCanonLayoutAlgorithm`)

nattress replied:
Ported both the `Array<T>` and USG layout and the tests pass 👍

nattress committed:
Port `Array<T>` and USG layout algorithms

Adds back another 5 passing tests

MichalStrehovsky commented on src/coreclr/src/tools/aot/ILCompiler.TypeSystem.ReadyToRun.Tests/TestTypeSystemContext.cs:
```diff
@@ -0,0 +1,136 @@
+
+        public override FieldLayoutAlgorithm GetLayoutAlgorithmForType(DefType type)
+        {
+#if !READYTORUN
```
Do we need to if READYTORUN anything in this file now that the dependencies are ported over?

nattress replied:
Gonzo

MichalStrehovsky reviewed (approved):
Looks good otherwise, thank you!

nattress committed:
Disable unit tests on linux-musl

On Alpine x64, ildasm executable cannot be found. The Ilasm SDK seems to be missing linux-musl recognition in its targets: https://github.com/dotnet/runtime/blob/54625f377552300ce03100b4d1689af30395f9fd/src/coreclr/src/.nuget/Microsoft.NET.Sdk.IL/targets/Microsoft.NET.Sdk.IL.targets#L32
